### ISSUE TITLE
Html Reader created invalid spreadsheets if <title> tag is too long

### ISSUE THREAD

**Issue by phinor** [PHPOffice/PhpSpreadsheet#729]
This is:.

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
When reading an HTML source file to create a spreadsheet, PhpSpreadsheet will truncate a lengthy (longer than 31 characters) <title> when setting the name of the current worksheet.

### What is the current behavior?
PhpSpreadsheet creates a file with a worksheet name that is too long, prompting Excel to perform "data rescue" operation on opening the file. The worksheet name is truncated by Excel.

### What are the steps to reproduce?
```php
<?php

require __DIR__ . '/vendor/autoload.php';

// Create HTML File
$html = "<html><head><title>1234567890123456789012345678901234567890</title></head><body>Content</body></html>";
file_put_contents ("temp.html", $html);

// Create HTML Reader
$objHTMLReader = IOFactory::createReader ("Html");
$excel = $objHTMLReader->load ("temp.html");

// Output File
header ('Content-Type: application/vnd.ms-excel');
header ('Content-Disposition: attachment;filename="Test.xlsx"');
$objWriter = IOFactory::createWriter ($excel, "Xlsx");
$objWriter->save ('php://output');
```

When opening the file produced, a warning message is displayed: "We found problems with some content in 'Test.xlsx'. Do you want us to try to recover as much as we can? If you trust the source of this workbook, click 'Yes'."

The Worksheet name is truncated to "123456789012345678901".

### Which versions of PhpSpreadsheet and PHP are affected?
PHP 7.2, PhpSpreadsheet 1.4.0

**Comment by kifni41** [comment]
I check the 'bug' and it's confirmed.
I tried to the sheet title on Libre Office, and it's seem that they support longer characters. I haven't check it on MS Office.

The title length limitation in on this _Worksheet/Worksheet.php_ in _setTitle_ function

I'm sure this is intended by anyone that code this function back then, but i'm not sure if this limitation is no longer needed now. 

Can anyone with deeper understanding about the spreadsheet reference confirm this matter?


```php
    public function setTitle($pValue, $updateFormulaCellReferences = true, $validate = true)
    {
        // Is this a 'rename' or not?
        if ($this->getTitle() == $pValue) {
            return $this;
        }

        // Old title
        $oldTitle = $this->getTitle();

        if ($validate) {
            // Syntax check
            self::checkSheetTitle($pValue);

            if ($this->parent) {
                // Is there already such sheet name?
                if ($this->parent->sheetNameExists($pValue)) {
                    // Use name, but append with lowest possible integer

                    if (Shared\StringHelper::countCharacters($pValue) > 29) {
                        $pValue = Shared\StringHelper::substring($pValue, 0, 29);
                    }
                    $i = 1;
                    while ($this->parent->sheetNameExists($pValue . ' ' . $i)) {
                        ++$i;
                        if ($i == 10) {
                            if (Shared\StringHelper::countCharacters($pValue) > 28) {
                                $pValue = Shared\StringHelper::substring($pValue, 0, 28);
                            }
                        } elseif ($i == 100) {
                            if (Shared\StringHelper::countCharacters($pValue) > 27) {
                                $pValue = Shared\StringHelper::substring($pValue, 0, 27);
                            }
                        }
                    }

                    $pValue .= " $i";
                }
            }
        }

        // Set title
        $this->title = $pValue;
        $this->dirty = true;

        if ($this->parent && $this->parent->getCalculationEngine()) {
            // New title
            $newTitle = $this->getTitle();
            $this->parent->getCalculationEngine()
                ->renameCalculationCacheForWorksheet($oldTitle, $newTitle);
            if ($updateFormulaCellReferences) {
                ReferenceHelper::getInstance()->updateNamedFormulas($this->parent, $oldTitle, $newTitle);
            }
        }

        return $this;
    }
```

**Comment by kifni41** [comment]
i just tried to trace the bug, it's seem that Worksheet/Worksheet.php failed catch some condition when trying to trim the title to 30.

This is the current walkaround to fix it, 
Just add 3 lines of code before if($validate)

not sure this is the correct way to do this. Let me know if this temporary fix can works for you.

```php
public function setTitle($pValue, $updateFormulaCellReferences = true, $validate = true)
{
    // Is this a 'rename' or not?
    if ($this->getTitle() == $pValue) {
        return $this;
    }

    // Old title
    $oldTitle = $this->getTitle();
   
   //Fix ? limiting title
    if (Shared\StringHelper::countCharacters($pValue) > 30) {
        $pValue = Shared\StringHelper::substring($pValue, 0, 30);
    }
    if ($validate) { ... ...
```

**Comment by oleibman** [comment]
Fixed by PR #3845 in Dec. 2023.
